LLM Request Migration on Engine Failure (#25)
ai-dynamo/enhancements: deps/NNNN-llm-request-migration.md

* Requires additional steps on the users to restart their requests; and
* Wastes the compute resource that generated the response up to the failure point; and
* Takes a longer time overall for the users to get their response

Ultimately leading to poor user experience.

## Goals

* A request migrated should continue the response from where it failed.
* A migration should be seamless so that the user is unlikely to notice any difference.

### Non Goals

* Prevent faults.
* Improve performance.

## Requirements

N/A

# Proposal

## Overview

The LLM Request Migration feature introduces a retry mechanism at the LLM Frontend that allows
requests to continue processing on different engine workers when connection failures occur. The
solution implements a configurable migration limit and tracks partial responses to enable seamless
continuation of generation from the point of failure.

## Architecture Components

### 1. Migration Limit Configuration

A new `--migration-limit` command-line flag is added across all engine types (VLLM, SGLang,
TensorRT-LLM) that specifies the maximum number of times a request can be migrated to another
worker. This parameter is:

> - Configurable per worker with a default value of 0 (no migration)

Reviewer: ```suggestion
- Configurable per model registration with a default value of 0 (no migration)
```

- Validated to be used only on worker nodes, not ingress nodes
- Propagated through the entire pipeline from launch flags to the migration operator

**Design Considerations:**

- **Engine Override Capability**: The design allows for future engine implementations to override
  the user-specified migration limit. This is important because certain engines or models may not
  generate properly with the current migration mechanism, requiring the engine to disable or reduce
  migration attempts to ensure correctness.

- **Multi-Model Flexibility**: Since a frontend may serve multiple models, setting the migration
  limit at the engine level provides users with the flexibility to configure different migration
  limits for different models based on their specific characteristics and reliability requirements.

### 2. Migration Operator

> A new `Migration` operator is introduced in the LLM processing pipeline that:

Reviewer: ```suggestion
A new `Migration` pipeline stage is introduced in the LLM processing pipeline that:
```

> - Sits between the Backend operator and the service backend in the processing chain

Reviewer: ```suggestion
- Sits between the Backend operator and the service backend in the processing chain
```
This is a little confusing - between the backend stage the engine client stage?

- **Retry Logic**:
  1. RetryManager detects NatsNoResponders error during initial stream creation
  2. Decrements remaining migration count and attempts new worker selection
  3. Creates fresh stream with original request (no partial state to preserve)
  4. Successfully processes all responses on retry
- **Key Insight**: Original request state is preserved since no generation has occurred yet

#### 3. Ongoing Request Migration (Mid-Stream Disconnection)
- **Scenario**: Connection lost during active generation after partial responses received
- **Failure Pattern**: "Stream ended before generation completed" error mid-stream
- **Retry Logic**:
  1. First stream successfully delivers partial responses
  2. RetryManager detects stream disconnection error and captures partial state
  3. Updates original request's token_ids vector with received tokens
  4. Creates new stream with accumulated context
  5. New worker continues generation from where the previous worker left off
  6. Delivers remaining responses to complete the generation
- **Key Insight**: Request state accumulation enables seamless continuation without restart

#### Token State Tracking

> The migration system maintains request continuity by:

Author: sure, added to the proposal

ticket tracking this: DIS-402
